src/installer.ts

```diff
@@ -1,38 +1,181 @@
   }
-  if (!tool) {
-    throw new Error('toolName parameter is required');
+
+  if (stackSetupGhc && stackVersion === '') {
+    errors.push('stack-version is required if stack-setup-ghc is set');
   }
-  if (!version) {
-    throw new Error('versionSpec parameter is required');
+
+  if (errors.length > 0) {
+    throw new Error(errors.join('\n'));
   }
 
-  const toolPath: string = path.join(baseInstallDir, tool, version, 'bin');
-  if (fs.existsSync(toolPath)) {
-    core.debug(`Found tool in cache ${tool} ${version}`);
-    core.addPath(toolPath);
+  return {
+    ghc: {
+      version: core.getInput('ghc-version') || def.ghc.version,
+      enable: !stackNoGlobal,
+      install: installGHC
+    },
+    cabal: {
+      version: core.getInput('cabal-version') || def.cabal.version,
+      enable: !stackNoGlobal,
+      install: installCabal
+    },
+    stack: {
+      version: stackVersion,
+      enable: stackVersion !== '',
+      install: installStack,
+      setup: core.getInput('stack-setup-ghc') !== ''
+    }
+  };
+}
+
+export const installCabal = async (version: string): Promise<void> =>
+  installTool('cabal', version);
+
+export const installGHC = async (version: string): Promise<void> =>
+  installTool('ghc', version);
+
+export async function installStack(version: string): Promise<void> {
+  const info =
+    version === 'latest'
+      ? 'Installing the latest version'
+      : `Installing version ${version}`;
+  core.startGroup(`${info} of stack`);
+  const platformMap = ({
+    linux: 'linux-x86_64-static',
+    darwin: 'osx-x86_64',
+    win32: 'windows-x86_64'
+  } as unknown) as Record<NodeJS.Platform, string>;
+
+  const name = `stack-${version}-${platformMap[process.platform]}`;
+  const url =
+    version === 'latest'
+      ? `get.haskellstack.org/stable/${platformMap[process.platform]}`
+      : `github.com/commercialhaskell/stack/releases/download/v${version}/${name}`;
+
+  const stack = await tc.downloadTool(`https://${url}.tar.gz`);
+  const p = await tc.extractTar(stack);
+
+  // Less janky than figuring out how to ./p/*/stack. (Not by much)
+  const stackPath =
+    (await fs.readdir(p, {withFileTypes: true}))
+      .flatMap(d => (d.isDirectory() ? [d.name] : []))
+      .find(f => f.startsWith('stack')) ?? '';
+
+  const cachedTool = await tc.cacheDir(join(p, stackPath), 'stack', version);
+  core.addPath(cachedTool);
+
+  core.endGroup();
+}
+
+type Tool = 'cabal' | 'ghc';
+async function installTool(tool: Tool, version: string): Promise<void> {
+  core.startGroup(`Installing ${tool}`);
+  // Linux comes pre-installed with some versions of GHC and supports older
+  // versions through hvr's PPA.
+  if (process.platform === 'linux') {
+    // Cabal is installed to /opt/cabal/x.x but cabal's full version is X.X.Y.Z
+    const v = tool === 'cabal' ? version.slice(0, 3) : version;
+
+    const p = join('/opt', tool, v, 'bin');
+    const installed = await fs
+      .access(p)
+      .then(() => true)
```

Comment:
I think it would be nice to have a message in the build logs to let you know that the tool was found in the cache and maybe another message if it's not pointing to documentation about which tools are in the vm by default. That documentation might include something like your suggested policy of N versions of ghc, with instructions or links about how to contribute to https://github.com/actions/virtual-environments (e.g., in the case of a new release).

Reply:
This brings up a small but interesting question. Cabal by default is installed with version x.x on Linux (because that's what hvr's PPA uses). This means cabal v3.0.0.0 is installed in `/opt/cabal/3.0/bin` and cabal v2.4.0.1 is installed in `/opt/cabal/2.4/bin`. It's fine, but also an odd design choice.

I currently break backwards compatibility and force an exact Cabal version only for implementation simplicity and secretly "undo" that when checking to see if I can locate a pre-installed version of Cabal.

This will lead to a somewhat confusing info message where we let people know we located cabal version "2.4" when they entered "2.4.1.0", necessitating a second message explaining that this is intended behavior.

Do we want to do that? If we don't, it seems like we might want to fully support "semver-like" versions (or at least major.major and have the "latest" point release be automatically selected).
But if we do that, it makes sense to also support GHC as well. If we went that route, it'd be more important to make sure that the code deferred to what the github virtual-environment thought was the latest point release rather than what it actually was, lest something get out of sync and we see CI build times spike right after a GHC release.

Reply:
Ahh cabal, how fun.

I think my vote right now would be to support both major.major.minor.patch and major.major (maybe just for cabal). So `3.0.0.0` and `3.0` would both work. Info messages should say something like "Found cabal 3.0.0.0 in /opt/cabal/3.0/bin/".

Reply:
That's what I was thinking as well.

I also looked at the state of how the code was structured and disliked how it special cased the happy path for linux. It'll be messy to update things to add fallback install methods and have the right "just add to path" behavior for pre-installed versions. And it doesn't cleanly handle non-exact versions. So I've restructured the code quite a bit over the last day and I'm much happier with how it looks.

I'll have that up soon-ish and that should have more informative messages and a cleaner way forward to setting the path for pre-installed versions once windows and macOS get pre-installed versions of GHC. The version handling will also be more flexible.